### PR TITLE
Add `.makeRequest()` to bundle exports

### PR DESCRIPTION
**Pull request by codetheweb**
I think this API is a bit more straightforward than a "module service", although we can still call it that in the docs.

This should be able to do everything the previous module service could, let me know if it's missing anything.

### CODE REVIEW

File: docs/embedding.md [hunk 1]
```diff
@@ -2,4 +2,23 @@
...
+export default withEdgeSpec({
+  auth: "none",
+  methods: ["GET"],
+})((req) => {
+  // This is a thin wrapper over `await import()` to enable types, you can use `import * as shipsApi from "@acme/star-wars-ships` directly if you want.
+  const shipsApi = await loadBundle("@acme/star-wars-ships")
+
+  return shipsApi.makeRequest(req, {
+    // This is important so that the ships API sees the correct path.
+    removePathnamePrefix: "/ships",
```
Comment (codetheweb) [inline docs/embedding.md]: open to other names for this

File: docs/embedding.md [hunk 2]
```diff
@@ -2,4 +2,23 @@
...
+  })
```
Comment (codetheweb) [inline docs/embedding.md]: will also be able to add custom middlewares here eventually

File: docs/embedding.md (hunk 1 above)
Comment (mxsdev) [reply docs/embedding.md]: I thought when I did the module service implementation that this is corrected automatically

i.e. that you only see the route path relative to the module service, instead of from root

Comment (mxsdev) [reply docs/embedding.md]: I think it would be preferable to have it resolved automatically if possible instead of making people provide the path manually

File: src/bundle/construct-manifest.ts [hunk 3]
```diff
@@ -44,7 +45,8 @@ const routeMapWithHandlers = {
 
 const edgeSpec = {
   routeMatcher: getRouteMatcher(Object.keys(routeMapWithHandlers)),
-  routeMapWithHandlers
+  routeMapWithHandlers,
+  makeRequest: async (req, options) => handleRequestWithEdgeSpec(edgeSpec, options)(req)
```
Comment (mxsdev) [inline src/bundle/construct-manifest.ts]: Could this wrap around the existing module service code? The code there is slightly different than a normal request, and for instance handles route parameter identification automatically

File: docs/embedding.md (hunk 1 above)
Comment (codetheweb) [reply docs/embedding.md]: I was thinking of an edge case where resolving automatically broke and you would have to provide the base path but thinking through it again I'm not sure there's a valid edge case

Comment (mxsdev) [reply docs/embedding.md]: Hmm, interesting. See the implementation of module services and let me know if you see anything fuzzy there.

I think we can provide something like this (`removePathnamePrefix`) later for sure though for custom use cases

File: src/bundle/construct-manifest.ts (hunk 3 above)
Comment (codetheweb) [reply src/bundle/construct-manifest.ts]: 👍 will try to add base path identification (I think that's what you mean?)

I figured that all adapters should be sharing the same logic here (want to allow Node/Bun adapters to remove the base path too) and imo a single function call is slightly easier than having to construct a new instance

Comment (mxsdev) [reply src/bundle/construct-manifest.ts]: > (I think that's what you mean?)

yeah sorry for unclear

> I figured that all adapters should be sharing the same logic here (want to allow Node/Bun adapters to remove the base path too) and imo a single function call is slightly easier than having to construct a new instance

yeah that's reasonable, I just wanna make sure that that code i wrote doesn't go to waste

you should probably also remove the old module service code too while you're at it, rather than just the tests

File: docs/embedding.md (hunk 1 above)
Comment (codetheweb) [reply docs/embedding.md]: ok edge case: your module service has a `[...path]/*` route at the root
I think we can still attempt to detect the base path by default though and allow overriding with a flag

Comment (mxsdev) [reply docs/embedding.md]: Why is that an edge case? Can't the `[...path]` merge with the `[...path]` of the parent?

Like certainly you could have

- Parent: `/path/to/module/[...path]`
- Child route `[...path]`

Then the child route just matches the parent route, right?

Comment (codetheweb) [reply docs/embedding.md]: yeah that should work

another one: your service has both `/health` and `/nested/health`, and is hosted at `/foo`

`/foo/nested/health` should be able to resolve to the correct path as long as your matching is greedy but if you misspell and fetch `/foo/neste/health` instead it hits `/health` instead of 404'ing

Comment (mxsdev) [reply docs/embedding.md]: no, that's not the way i implemented it for module services

for module services it substitutes so that the spread route parameter becomes the child's route

So in that case it would still 404 because the parent is `/foo/[...path]`, so the route becomes `/neste/health`, which doesn't exist in the module service causing a 404

Comment (codetheweb) [reply docs/embedding.md]: that assumes the parent application is using edgespec/has `.routeParams` available then? 

tbh I'm having a bit of a hard time understanding your implementation

sorry I can sense the frustration lol

Comment (codetheweb) [reply docs/embedding.md]: e.x. a simple greedy matcher that works in most cases:

Screenshot 2024-02-14 at 11 48 06

Comment (codetheweb) [reply docs/embedding.md]: ^ example implementation in 3ec5cb4

**Review by mxsdev — APPROVED** [review]
looks good, maybe just update docs?
